src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -445,51 +492,154 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
-                    // consumes target ref, data ptr and size, pushes nothing
+                    // Consumes target ref, data ptr and size, pushes nothing.
                     _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: -3);
+                    if (used)
+                    {
+                        EmitExpression(inPlaceTarget, used: true);
+                    }
                 }
                 else
                 {
-                    // consumes data ptr and size, pushes the instance
+                    // Consumes data ptr and size, pushes the instance.
                     _builder.EmitOpCode(ILOpCode.Newobj, stackAdjustment: -1);
                 }
 
-                EmitSymbolToken(ctor.AsMember(spanType), wrappedExpression.Syntax, optArgList: null);
+                EmitSymbolToken(rosPointerCtor.AsMember(spanType), wrappedExpression.Syntax, optArgList: null);
+                return true;
             }
 
+            // We're dealing with a primitive that's larger than a single byte.
+            Debug.Assert(specialElementType.SizeInBytes() is 2 or 4 or 8, "Supported primitives are expected to be 2, 4, or 8 bytes");
+
+            if (lengthForConstructor != elementCount)
+            {
+                // We need to use RuntimeHelpers.CreateSpan / cached array, but the code has requested a subset of the elements.
+                // That means the code is something like `new ReadOnlySpan<char>(new[] { 'a', 'b', 'c' }, 1, 2)`
+                // rather than `new ReadOnlySpan<char>(new[] { 'b', 'c' })`.  If such a pattern is found to be
+                // common, this could be augmented to accomodate it.  For now, we just return false to fail
+                // to optimize this case.
+                return false;
+            }
+
+            if (inPlaceTarget is not null)
+            {
+                // We can use RuntimeHelpers.CreateSpan / cached array but not for in-place initialization. Fail to optimize,
+                // but tell the caller they can call this again with a null inPlaceTarget, at which point this
+                // should be able to optimize the call.
+                avoidInPlace = true;
+                return false;
+            }
+
+            // As we're dealing with multi-byte types, endianness needs to be considered. Such handling is provided by the
+            // runtime's RuntimeHelpers.CreateSpan, which will wrap a span around the blob on little endian and which will
+            // allocate an array and cache it on big endian.
+            MethodSymbol? createSpan = (MethodSymbol?)_module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);
+            if (createSpan is not null)
+            {
+                if (createSpan.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error)
+                {
+                    // Something went wrong.
+                    return false;
```

Comment:
Were you able to hit this case? (an Obsolete attribute will probably do it)

Reply:
I wasn't able to trigger this.  I tried putting an ObsoleteAttribute on CreateSpan, even with error: true, and it didn't hit this path.

Reply:
> I wasn't able to trigger this.  I tried putting an ObsoleteAttribute on CreateSpan, even with error: true, and it didn't hit this path.

See unit test ```StaticFieldIsUsedForSpanCreatedFromArrayWithInitializer_02``` that covers scenario for the ```rosPointerCtor```

Reply:
Doesn't that test use MakeMemberMissing to make it so that GetWellKnownTypeMember returns null rather than there being an error at the use site?  Or have I misunderstood?

Reply:
It also covers scenario when method is there and would be found by compilation (as you do that right now), but is bad and should not be used, Read the comment at the beginning of the unit-test.

Reply:
It also covers scenario when method is there and would be found by compilation (as you do that right now), but is bad and should not be used, Read the comment at the beginning of the unit-test.